shell/platform/darwin/macos/framework/Source/FlutterEmbedderKeyResponder.mm

```diff
@@ -0,0 +1,715 @@
+ *
+ * Examples of unprintable keys are "NSUpArrowFunctionKey = 0xF700" or
+ * "NSHomeFunctionKey = 0xF729".
+ *
+ * See
+ * https://developer.apple.com/documentation/appkit/1535851-function-key_unicodes?language=objc
+ * for more information.
+ */
+static uint64_t KeyOfPlane(uint64_t baseKey, uint64_t plane) {
+  return plane | (baseKey & kValueMask);
+}
+
+/**
+ * Returns the physical key for a key code.
+ */
+static uint64_t GetPhysicalKeyForKeyCode(unsigned short keyCode) {
+  NSNumber* physicalKey = [keyCodeToPhysicalKey objectForKey:@(keyCode)];
+  if (physicalKey == nil)
+    return 0;
+  return physicalKey.unsignedLongLongValue;
+}
+
+/**
+ * Returns the logical key for a modifier physical key.
+ */
+static uint64_t GetLogicalKeyForModifier(unsigned short keyCode, uint64_t hidCode) {
+  NSNumber* fromKeyCode = [keyCodeToLogicalKey objectForKey:@(keyCode)];
+  if (fromKeyCode != nil)
+    return fromKeyCode.unsignedLongLongValue;
+  return KeyOfPlane(hidCode, kHidPlane);
+}
+
+/**
+ * Converts upper letters to lower letters in ASCII, and returns as-is
+ * otherwise.
+ *
+ * Independent of locale.
+ */
+static uint64_t toLower(uint64_t ch) {
+  const uint64_t lowerA = 0x61;
+  const uint64_t upperA = 0x41;
+  const uint64_t upperZ = 0x5a;
+  if (ch >= upperA && ch <= upperZ)
+    return ch + (lowerA - upperA);
+  return ch;
+}
+
+/**
+ * Returns the logical key of a KeyUp or KeyDown event.
+ *
+ * For FlagsChanged event, use GetLogicalKeyForModifier.
+ */
+static uint64_t GetLogicalKeyForEvent(NSEvent* event, uint64_t physicalKey) {
+  // Look to see if the keyCode can be mapped from keycode.
+  NSNumber* fromKeyCode = [keyCodeToLogicalKey objectForKey:@(event.keyCode)];
+  if (fromKeyCode != nil)
+    return fromKeyCode.unsignedLongLongValue;
+
+  NSString* keyLabel = event.charactersIgnoringModifiers;
+  NSUInteger keyLabelLength = [keyLabel length];
+  // If this key is printable, generate the logical key from its Unicode
+  // value. Control keys such as ESC, CTRL, and SHIFT are not printable. HOME,
+  // DEL, arrow keys, and function keys are considered modifier function keys,
+  // which generate invalid Unicode scalar values.
+  if (keyLabelLength != 0 && !IsControlCharacter(keyLabelLength, keyLabel) &&
+      !IsUnprintableKey(keyLabelLength, keyLabel)) {
+    // Given that charactersIgnoringModifiers can contain a string of arbitrary
+    // length, limit to a maximum of two Unicode scalar values. It is unlikely
+    // that a keyboard would produce a code point bigger than 32 bits, but it is
+    // still worth defending against this case.
+    NSCAssert((keyLabelLength < 2), @"Unexpected long key label: |%@|.", keyLabel);
+
+    uint64_t codeUnit = (uint64_t)[keyLabel characterAtIndex:0];
+    if (keyLabelLength == 2) {
+      uint64_t secondCode = (uint64_t)[keyLabel characterAtIndex:1];
+      codeUnit = (codeUnit << 16) | secondCode;
+    }
+    if (codeUnit < 256) {
```

Comment:
I'm not sure this conditional is necessary.  Sure, you save a function call for higher Unicode characters, but the compiler is likely to inline the function anyhow, and the first thing it does is check the range.

Probably just better to eliminate this entirely, and change the next line to:
```c
return KeyOfPlane(toLower(codeUnit), kUnicodePlane);
```